### ISSUE TITLE
Matching the geometry to the input image

### ISSUE THREAD

**Issue by jamalknight** [facebookresearch/meshrcnn#22]
Hi there

I have a question about aligning the geometry to a camera in a 3D app like Maya.  
Object classification/segmentation is ok [example](https://github.com/jamalknight/misc/blob/master/bunS1_01_bg_v001.0001.jpg)

The geometry obj file is created, but I was wondering if there is a way to align the geometry with a perspective camera that matches the original image. 
 - This geometry is in 3d space not aligned [here](https://github.com/jamalknight/misc/blob/master/bunS1_01_geoScene.0001.jpg)
 - The geometry is manually aligned here by eye [here](https://github.com/jamalknight/misc/blob/master/bunS1_01_3d_v001.0001.jpg), but might not be accurate

I would like it to match the image - is there a relatively simple way this could be done?

**Comment by gkioxari** [comment]
Hi @jamalknight! This is an excellent question! I want to give a complete (and thus somewhat long!) response, as I think this will be useful for others as well. My answer below assumes a perspective camera.

### What does Mesh R-CNN output?
First, let's see what Mesh R-CNN outputs. Mesh R-CNN returns the 3D shape of an object in the camera coordinate system confined in a 3D box which respects the aspect ratio of the object detected in the image. If you provide the _focal length_ `f` of the camera and the actual depth location `t_z` of the object's center , i.e. how far the center of the object is from the image plane in the Z axis, then Mesh R-CNN would align the predicted 3D object shape with the image. Note that this assumes we know both `f` and `t_z`. Also note that Mesh R-CNN does not make a prediction for `t_z` because Pix3D does not contain the actual depth of the objects. In Pix3D, the tuple (`f`, `t_z`) provided by the annotations does not correspond to the actual camera metadata nor metric depth of the object but is computed subsequently at annotation time by their annotation process and annotation tool and thus is somewhat adhoc and doesn't correspond to actual values. This is the reason we don't tackle the problem of estimating `t_z` (this problem is also called the scene layout prediction problem). Finally, note that our demo assumes a value for `f` and `t_z` when making predictions.

### Depth-scale ambiguity from a single image
I assume that in most cases you have access to the focal length `f` of the camera from its metadata. Given `f`, there is now an infinite set of possible `(s, t_z)`, where `s` is the scale of the object and `t_z` is its center's depth (as discussed above). This is the so known depth-scale ambiguity, which means that the image can be explained by either placing a smaller scale object close to the image plane or a larger scale object further away from the image plane (see attached image). Depth-scale ambiguity is an inherent issue when making predictions from a monocular image. 

### Silhouette based rendering to the rescue!
Predicting the exact position and scale of an object from a monocular image is difficult as discussed above. However, if you only care to align the 3D object with the image plane and you don't care to necessarily predict its true scale `s` or its true metric depth `t_z`, then there is a solution! You can fix `t_z` to be any metric depth value, e.g. 1m, and then you can estimate `s` by using a mesh renderer and aligning the 2D silhouette of the object (its 2D mask as predicted by Mesh R-CNN) with its rendered silhouette. This problem is called silhouette based rendering and in this setup it requires optimizing over a single scalar value `s`. For this you can use, [PyTorch3D](https://github.com/facebookresearch/pytorch3d) which contains a fast and modular differentiable renderer and optimize over `s` using simple optimization methods, e.g. SGD, to find `s`. There is no neural networks required for this, nor other learned parameters, just good old fashioned optimization. There is a [tutorial](https://github.com/facebookresearch/pytorch3d/blob/master/docs/tutorials/camera_position_optimization_with_differentiable_rendering.ipynb) doing something slightly different that should be helpful. In that tutorial, the parameters optimized are the camera position parameters. All you need to do is change the formulation and solve over the scale of the object `s` instead. 

Now if you end up doing something like this for your problem and want to contribute it back to Mesh R-CNN, I'd love to review your PR and incorporate it in our codebase! 

Ok that was super long but hopefully it was helpful!

**Comment by vadimkantorov** [comment]
@gkioxari Could you please share the blender script (if you still have it) you used to produce the third image?

**Comment by vadimkantorov** [comment]
I've imported the "chair" mesh (that I obtained from running demo.py) into Blender, but the camera looks away from the object. Do I need to manually reset the camera position / orientation? If yes, what should it be?

Thank you!

<img width="1792" alt="Screenshot 2021-08-06 at 11 03 34" src="https://user-images.githubusercontent.com/1041752/128486260-9dd7f6bd-3a3b-4c7e-a0a8-a26f6da4f645.png">

**Comment by vadimkantorov** [comment]
I am currently using (location: (0, 0, 0), rotation: (180deg, 0, 0), focal length: 20mm, clip start/end: 0.1m/100m)

I'm still getting a gray render when I go to camera view.

Where should I put the light?

Here's the original image and mesh produced by MeshRCNN: [chair3.zip](https://github.com/facebookresearch/meshrcnn/files/6945031/chair3.zip)

**Comment by vadimkantorov** [comment]
Or maybe the problem is somehow clipping...

**Comment by vadimkantorov** [comment]
Hmm, enabling "Depth of Field" makes it show something. Not sure why "Depth of Field" is needed...

**Comment by vadimkantorov** [comment]
Now I got the one below. Does it make sense?

![0_mask_chair_1 000](https://user-images.githubusercontent.com/1041752/128509296-ec57a44e-cef4-4e3b-8c71-c285dd54c872.png)

<img width="1791" alt="Screenshot 2021-08-06 at 14 17 52" src="https://user-images.githubusercontent.com/1041752/128509245-f9fe509d-614a-4edb-a6ca-78a9f4a9331e.png">
